src/line_launcher/mod.rs

```diff
@@ -141,6 +138,173 @@ pub fn get_combined_message_receiver(
     receiver
 }
 
+#[derive(Clone, PartialEq, Eq, Debug, Copy)]
+struct MeasureBeat {
+    beat_number: BeatNumber,
+    measure: u32,
+}
+
+impl MeasureBeat {
+    pub fn new(beat_number: BeatNumber, measure: u32) -> Self {
+        Self {
+            beat_number,
+            measure,
+        }
+    }
+
+    pub fn increment(&mut self) {
+        if self.beat_number.sixteenth_note == 15 {
+            self.measure += 1;
+        }
+        self.beat_number.increment();
+    }
+
+    pub fn incremented(&self) -> Self {
+        let mut cloned = *self;
+        cloned.increment();
+        cloned
+    }
+
+    pub fn incremented_by(&self, num_sixteenths: u32) -> Self {
+        let mut cloned = *self;
+        for _ in 0..num_sixteenths {
+            cloned.increment();
+        }
+        cloned
+    }
+}
+
+impl Default for MeasureBeat {
+    fn default() -> Self {
+        Self {
+            beat_number: BeatNumber::new(0),
+            measure: 0,
+        }
+    }
+}
+
+#[derive(Debug)]
+struct NoteAndMeasureBeat {
+    note: wmidi::Note,
+    measure_beat: MeasureBeat,
+}
+
+#[derive(Debug)]
+struct PlannedNote {
+    note_on: NoteAndMeasureBeat,
+    note_off: NoteAndMeasureBeat,
+    has_note_on_fired: bool,
+    has_note_off_fired: bool,
+}
+
+impl PlannedNote {
+    pub fn new(line_note: &LineNote, measure_beat_start: MeasureBeat, pitch_offset: i8) -> Self {
+        let note_adjusted = line_note.note.step(pitch_offset).unwrap();
+        Self {
+            note_on: NoteAndMeasureBeat {
+                note: note_adjusted,
+                measure_beat: measure_beat_start,
+            },
+            note_off: NoteAndMeasureBeat {
+                note: note_adjusted,
+                measure_beat: measure_beat_start.incremented_by(line_note.duration),
+            },
+            has_note_on_fired: false,
+            has_note_off_fired: false,
+        }
+    }
+}
+
+#[derive(Debug)]
+struct PlannedNotes {
+    planned_notes: Vec<PlannedNote>,
+}
+
+impl PlannedNotes {
+    pub fn push(&mut self, planned_note: PlannedNote) {
```

Comment:
There's some tension here where it feels correct to have a `PlannedNotes` abstraction but then most of what I want to do to it is directly manipulate its single `Vec` struct member - here I wrote out a wrapper around its `.push()` method but there are other places where I just "reach in" and do eg `planned_notes.planned_notes.something()`